src/internal/grpcutil/iter.go

```diff
@@ -0,0 +1,53 @@
+package grpcutil
+
+import (
+	"context"
+	"io"
+
+	"github.com/gogo/protobuf/proto"
+	"google.golang.org/grpc"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/stream"
+)
+
+type ClientStream[T proto.Message] interface {
+	Recv() (T, error)
+	grpc.ClientStream
+}
+
+type iterator[T proto.Message] struct {
+	cs ClientStream[T]
+}
+
+func newIterator[T proto.Message](cs ClientStream[T]) iterator[T] {
+	return iterator[T]{cs: cs}
+}
+
+func (it iterator[T]) Next(ctx context.Context, dst *T) error {
```

Comment:
I see the friendlier API below and this makes sense.  It's especially nice for the pre-allocated slice case.

For whatever reason I have never associated much cost with "return <pointer>" vs *result = pointer, but I don't have any actual evidence to back that up.

Reply:
Yeah this is not a performance boost.  This is just the most straightforward mapping of the gRPC stuff to the generic iterators.

The next steps would be experimenting with calling `Reset()` on a single gRPC object.  That would allow us to reuse a single gRPC and any of it's backing buffers for new data.